### DISCUSSION TITLE
When should I override OnCanExecuteAsync/OnCanExecute?

### DISCUSSION THREAD

**Discussion by e-mair** [elsa-workflows/elsa-core#3101] [Q&A]
Hello,

I am trying to understand the activity lifecycle in a bit more detail. I am creating a custom activity and overriding the `OnCanExecute` method to perform some business logic; check my inputs are not null etc. However, I never seem to be hitting a breakpoint in the method when I debug my activity.

Here is an example activity:
```
internal sealed class MyCustomActivity : Activity
{
     [ActivityInput(Hint = "Some important input.",
          UIHint = ActivityInputUIHints.SingleLine,
          SupportedSyntaxes = new[] { SyntaxNames.Literal, SyntaxNames.Liquid, SyntaxNames.JavaScript })]
     public int? SomeInput { get; set; }

     [ActivityOutput(Hint = "Some valuable information.")]
     public object? SomeOutput { get; set; }

     // Constructor goes here.

     protected override bool OnCanExecute(ActivityExecutionContext context) => SomeInput is not null;

     protected override async ValueTask<IActivityExecutionResult> OnExecuteAsync(ActivityExecutionContext context)
     {
          // Main activity workload goes here.    
     }
}
```

Will `OnCanExecuteAsync` or `OnCanExecute` be executed on _every_ execution attempt of the activity? Or is there some other override I should use to perform these checks?

**Comment by terry-delph** [comment]
I would also like to understand the lifecycle as I get unexpected results when using it.
It appears to only call `OnCanExecute `when returning from a suspended state or when the activity is the first in the workflow.
This means that if my second activity overrides the `OnCanExecute `method, then it never gets called unless itself has been suspended.

@sfmskywalker
What is the purpose of the `OnCanExecute `interface?
Is the definition misleading or possibly not fully implemented throughout the Workflow Runner lifecycle?
Should it activate on all activities, regardless of coming back from a suspended state?

Thanks

**Reply by samadpaydar3** [reply]
@terry-delph 

Hi,

Did you find out how the OnCanExecute works? Yes. It looks like it is called only for the first activity, but I think this is not what is expected. 

Thanks.
